### PR TITLE
Stabilize Self and associated types in struct expressions and patterns

### PR DESCRIPTION
**Pull request by petrochenkov**
Rebase of https://github.com/rust-lang/rust/pull/37734
Closes https://github.com/rust-lang/rust/issues/37544
r? @nikomatsakis

### CODE REVIEW

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -378,6 +374,8 @@ declare_features! (
     // Allows `..` in tuple (struct) patterns
     (accepted, dotdot_in_tuple_patterns, "1.14.0", Some(33627)),
     (accepted, item_like_imports, "1.14.0", Some(35120)),
+    // Allows using `Self` and associated types in struct expressions and patterns.
+    (accepted, more_struct_aliases, "1.14.0", Some(37544)),
```
Comment (bluss) [inline src/libsyntax/feature_gate.rs]: Great feature! does this version need to be updated (Should be 1.16, right)?

Comment (petrochenkov) [inline src/libsyntax/feature_gate.rs]: I don't know.
For `active` features it's the version in which it was implemented. What does this version mean exactly for `accepted` and `removed` features, something else?
I never cared to update it previously.

Comment (bluss) [inline src/libsyntax/feature_gate.rs]: I see, let's see if anyone can say if there's a reason to update it. For library features it's neat to see it in the docs, I guess here it doesn't actually show anywhere?

**Comment by KalitaAlexey** [comment]
@petrochenkov,
Has the documentation been updated?
Where can I read about places where I can use `Self`?

**Comment by petrochenkov** [comment]
@KalitaAlexey 
>Where can I read about places where I can use Self?

Now - everywhere!
https://doc.rust-lang.org/reference.html#self-types looks up-to-date.
If people actually read these docs, then I'd expect they to write `Self { .... }` intuitively and then be surprised about the error. One extra example with `Self` in a struct literal wouldn't hurt though.

File: src/libsyntax/feature_gate.rs (hunk 1 above)
Comment (zackmdavis) [inline src/libsyntax/feature_gate.rs]: > What does this version mean exactly for accepted and removed features, something else?

Note [the comment](https://github.com/rust-lang/rust/blob/d7777ae6/src/libsyntax/feature_gate.rs#L96-L97) above the first `declare_features!` block: "The version numbers here correspond to the version in which the current status was set."  

> I guess here it doesn't actually show anywhere?

It is at least [displayed in the message of the lint](https://github.com/rust-lang/rust/pull/33973) for `#[feature]`-gate attributes for already-stable features.

Comment (zackmdavis) [inline src/libsyntax/feature_gate.rs]: PR: #39552

**Pull request merged**